### ISSUE TITLE
CustomLayer + three.js // need for camera position

### ISSUE THREAD

**Issue by indus** [mapbox/mapbox-gl-js#7395]
**mapbox-gl-js version**: v0.50.0-beta.1

Hi I'm experimenting with the new CustomLayer and using three.js to augment the map. I'm now facing multiple situations (raycasting, dynamic shader, etc) where I would need the camera position & rotation (its transformation matrix). If I understand the projectionMatrix (that gets feed to the render function) correct it is somehow included in there. But I don't know how to retrieve it from the 4x4 matrix (if it is even possible).

Do you have a solution for this problem?

It seems to me that interoperability between MB-GL-JS and three.js is a bit limited because you are working with a complete projection matrix while three.js keeps a reduced projectionMatrix and a viewMatrix sepereated and multiply them in the shader.  Is this correct?

**Comment by ansis** [comment]
Thanks for the feedback @indus !

Would you want the camera position and rotation as separate values or a transformation matrix that incorporates them both? So [position](https://threejs.org/docs/#api/en/core/Object3D.position) would be the mercator coordinates of the camera and [rotation](https://threejs.org/docs/#api/en/core/Object3D.rotation) would be a euler angle? Is there anything else you would need that we aren't providing?

> Do you have a solution for this problem?

Not yet, but we should. We don't currently store the position of the camera anywhere because we're specifying the camera as what it's looking at instead of where it's looking from. https://github.com/mapbox/mapbox-gl-js/pull/6093 implements a way to convert that to a position.

Is there anything else that we should expose or provide?

**Comment by ryanhamley** [comment]
If #6093 is useful for this, I'd be happy to pick the PR up and get it merged. Should those methods be exposed on `camera`/`map`?

**Comment by indus** [comment]
@ansis It is hard to say what would be **most** usefull but I've looked into the code a bit and I would guess when it comes to three.js a seperateion of the [perspective](https://github.com/mapbox/mapbox-gl-js/blob/master/src/geo/transform.js#L568) and the [transformation](https://github.com/mapbox/mapbox-gl-js/blob/master/src/geo/transform.js#L570-L583) would be nice.

In three.js they get multiplied in the shader: https://github.com/mrdoob/three.js/blob/34dc2478c684066257e4e39351731a93c6107ef5/examples/js/shaders/BasicShader.js#L15

I mean you don't have to do it like three.js does, but it seems pretty common to have three separate matrices (model-, view- and projection-matrix): http://www.opengl-tutorial.org/beginners-tutorials/tutorial-3-matrices/#the-model-view-and-projection-matrices

According to this article what you are returning in the render function (and storing in the map.transform) isn't a projectionMatrix but a viewProjectionMatrix.

If you would provide them seperated it would be pretty easy to get the position and the rotation of the camera (from the viewMatrix).

Right now it is not so easy, and until now I haven't found a way how to do it. An idea I have right now is to multiply the inverse projectionMatrix (like [so](https://github.com/mapbox/mapbox-gl-js/blob/master/src/geo/transform.js#L568); the values are in map.transform) to the provided viewProjectionMatrix to get the viewMatrix (just an idea - I'm not good with this stuff ;-)

**Comment by indus** [comment]
@ryanhamley while the feature proposed in the PR sounds helpful in general and would bring a reasonable addition. I would guess that it won't be a big benefit for the sort of implementation problems I tried to describe. In my case I'm not dealing with coordinates but only gl-units (there would be additional transformation steps involved that require more helper functions).
In my understanding the splendid CustomLayer implementation should be pretty low-level but easily accessible for WebGL enthusiasts.

**Comment by indus** [comment]
I had success with decoupling the provided ```viewProjectionMatrix``` into a ```projectionMatrix``` and a ```viewMatrix```. Three.js also provides a pleasent "decompose" function to get position, rotation and scale from the ```viewMatrix```.

Thats the code:

``` TS
render(gl, viewProjectionMatrix) {

    const transform = this.map.transform;
    const camera = this.camera;

    const projectionMatrix = new Float64Array(16);
    const projectionMatrixI = new Float64Array(16);
    const viewMatrix = new Float64Array(16);
    const viewMatrixI = new Float64Array(16);

    // https://github.com/mapbox/mapbox-gl-js/blob/master/src/geo/transform.js#L556-L568
    const halfFov = transform._fov / 2;
    const groundAngle = Math.PI / 2 + transform._pitch;
    const topHalfSurfaceDistance = Math.sin(halfFov) * transform.cameraToCenterDistance / Math.sin(Math.PI - groundAngle - halfFov);
    const furthestDistance = Math.cos(Math.PI / 2 - transform._pitch) * topHalfSurfaceDistance + transform.cameraToCenterDistance;
    const farZ = furthestDistance * 1.01;

    mat4.perspective(projectionMatrix, transform._fov, transform.width / transform.height, 1, farZ);
    mat4.invert(projectionMatrixI, projectionMatrix);
    mat4.multiply(viewMatrix, projectionMatrixI, viewProjectionMatrix);
    mat4.invert(viewMatrixI, viewMatrix);

    camera.projectionMatrix = new THREE.Matrix4().fromArray(<any>projectionMatrix);
    camera.matrix = new THREE.Matrix4().fromArray(<any>viewMatrixI);

    // camera.matrix.decompose(camera.position, camera.quaternion, camera.scale);
    // console.log(camera.position)

    this.renderer.state.reset();
    this.renderer.render(this.scene, this.camera);
  }
```

the code to calculate ```projectionMatrix``` and ```projectionMatrixI``` could propably be moved out of the render function and only be recalculated when the pitch or the screen size changes.

After all I would still prefer to get ```projectionMatrix``` and ```viewMatrix``` passed in to the render function as two seperated values.

**Comment by indus** [comment]
While producing the same viewProjectionMatrix it seems that the viewMatrix and the projectionMatrix have been somehow inverted. The following now works for me:
``` TS
    mat4.perspective(projectionMatrix, transform._fov, transform.width / transform.height, 1, farZ);
    mat4.invert(projectionMatrixI, projectionMatrix);
    mat4.multiply(viewMatrix, projectionMatrixI, viewProjectionMatrix);
    mat4.multiply(projectionMatrix, viewProjectionMatrix, viewMatrix)

    camera.projectionMatrix = new THREE.Matrix4().fromArray(<any>projectionMatrix);
    camera.matrix = new THREE.Matrix4().fromArray(<any>viewMatrix);
    camera.matrix.decompose(camera.position, camera.quaternion, camera.scale);
```
As I said I'm not very good with this stuff.

**Comment by robhawkes** [comment]
Could this be related to why I'm unable to get the `THREE.EffectComposer` to work correctly with the new custom layers functionality? As soon as I utilise the composer the depth information seems to be lost and the Three.js objects appear on top of the Mapbox objects.

Example with a simple `THREE.RenderPass`: https://jsfiddle.net/robhawkes/bn49y0oj/
Example using `THREE.UnrealBloomPass` (doesn't even show up): https://jsfiddle.net/robhawkes/zvjofaxw/

I'm hesitant to open an issue here about it as I'm not sure it's a bug with Mapbox per se, though I'm unsure what the fix is (whether with my Three.js code or otherwise).

Happy to move this to a separate issue if appropriate.

**Comment by sansumbrella** [comment]
You might be interested in trying out [threebox](https://github.com/peterqliu/threebox/blob/master/src/Camera/CameraSync.js), which provides helpers for doing these camera projections/synchronization with Three.js.

**Comment by frading** [comment]
Thanks a lot to @indus and Threebox to try and provide solutions to this problem. I'd really need a camera position as well to do some raymarching and display volumetric shaders over a map.

A problem I am facing at the moment is that @indus's solution creates a camera that is scaled with very small values (0.00006103515625 in my current test) and an axis scaled in the negatives. This makes handling rays in glsl trickier than it already is.

I tried to un-scale the camera, but this impacts the near clipping plane and objects disappear very quickly. I then tried to provide a different `near` than 1 to the call `mat4.perspective(projectionMatrix, transform._fov, transform.width / transform.height, near, farZ)`, but this seems to impact the position of the camera.

I also tried to use Threebox, but it comes bundled with its own version of threejs, so doesn't make it easy to integrate with an existing project.

Also both of those solutions seem to use code extracted from mapbox libraries, as well as `map.transform`, which isn't documented, so it feels like this could change anytime.

If that adds anything to the conversation, I could somehow get some raymarching working when displaying a cube as large as a continent. But when working at the street level, the raymarching became unstable.

In short, at the moment, my understanding of projection matrices isn't deep enough to properly solve the problem. So it would be great if there was a stable way to get a non-scaled camera, with a projection matrix as well as a transform one.

**Comment by ansis** [comment]
@frading thanks for the feedback! This is something we're interested in fixing. I'll make sure to pull you in on reviewing any work on this. If possible, a full example of working content-sized cube or the broken street-level cube would be helpful.

**Comment by frading** [comment]
Thanks for your reply! And yes, I'd love to provide you with an example. There are lots of moving parts to extract but I'll give it a go.

**Comment by yuanzhaokang** [comment]
> If someone else have trouble getting the Three js Raycaster to work with custom layers, I've used this code successfully:
> 
> ```ts
>     itemsAtPosition(pos: [number, number]) {
> 
>         if (!this.cursor)
>             return;
> 
>         const cameraPosition = new THREE.Vector3(0, 0, 0)
>             .unproject(this.camera)
>             ;
> 
>         const mousePos = new THREE.Vector3(pos[0], pos[1], 0.99)
>             .unproject(this.camera)
>             ;
> 
>         const direction = mousePos
>             .clone()
>             .sub(cameraPosition)
>             .normalize()
>             ;
> 
>         this.raycaster.near = -1;
>         this.raycaster.far = 5;
>         this.raycaster.ray.set(mousePos, direction);
> 
>         return this
>             .raycaster
>             .intersectObjects(this.scene.children)
>             .map(o => o.object)
>             ;
>     }
> 
>     // with pos calculated like this:
> 
>     const rect = this.mapContainer.getBoundingClientRect();
> 
>     const pos: [number, number] = [
>         -1 + 2 * (e.clientX - rect.left) / this.mapContainer.clientWidth,
>         1 - 2 * (e.clientY - rect.top) / this.mapContainer.clientHeight
>     ];
> ```

@markusjohnsson 
hi, could you please show all the code of your example? Thanks.

**Comment by markusjohnsson** [comment]
@yuanzhaokang not really as it is part of a larger proprietary application. I've updated the code with some missing parts. I can elaborate further if you have problems.

**Comment by yuanzhaokang** [comment]
> @yuanzhaokang not really as it is part of a larger proprietary application. I've updated the code with some missing parts. I can elaborate further if you have problems.

@markusjohnsson Thanks very much.

**Comment by peterqliu** [comment]
@yuanzhaokang check out threebox for functional raycasting out of the box. [Here's](https://bl.ocks.org/peterqliu/c55d5d0548f7b17ee01e7b23c007c06e) an example of it to highlight a cube on hover

**Comment by yuanzhaokang** [comment]
> @yuanzhaokang check out threebox for functional raycasting out of the box. [Here's](https://bl.ocks.org/peterqliu/c55d5d0548f7b17ee01e7b23c007c06e) an example of it to highlight a cube on hover

@peterqliu Thanks. But threebox maybe lost the depth test of the cube.  
![top](https://user-images.githubusercontent.com/16620101/48597357-52e76f80-e998-11e8-9e11-6e7e28776da8.jpg)
![front](https://user-images.githubusercontent.com/16620101/48597376-698dc680-e998-11e8-8ea1-414f4459c7d1.jpg)

**Comment by peterqliu** [comment]
@yuanzhaokang that is the intended behavior -- the custom layer draws within the context of the 2D layer stack, regardless of whether the layers themselves are 3D. For proper depth testing, you would put all 3D elements into the same custom layer

**Comment by yuanzhaokang** [comment]
> @yuanzhaokang that is the intended behavior -- the custom layer draws within the context of the 2D layer stack, regardless of whether the layers themselves are 3D. For proper depth testing, you would put all 3D elements into the same custom layer

@peterqliu Thanks. Give me a lot of help !!!

**Comment by Kamrad** [comment]
> that is the intended behavior -- the custom layer draws within the context of the 2D layer stack, regardless of whether the layers themselves are 3D. For proper depth testing, you would put all 3D elements into the same custom layer
@peterqliu Is it possible to set it to use the same depth space? Like in an example [here ](https://github.com/mapbox/mapbox-gl-js/pull/7039).

**Comment by cc-ch** [comment]
> > @yuanzhaokang not really as it is part of a larger proprietary application. I've updated the code with some missing parts. I can elaborate further if you have problems.
>  
> @markusjohnsson Thanks very much.



> > @yuanzhaokang that is the intended behavior -- the custom layer draws within the context of the 2D layer stack, regardless of whether the layers themselves are 3D. For proper depth testing, you would put all 3D elements into the same custom layer
> 
> @peterqliu Thanks. Give me a lot of help !!!

Did you use the above code to solve the problem? I have used the customer layer by three.js, the Raycaster have some problem. The three.js's camera can not follow the mapbox camera. Can u show me some code about it. thx

**Comment by mjcc007** [comment]
I use the babylon as the webgl 3d engine and create a custom layer but i got the error in the callback function render "camera.js?a778:37 Uncaught TypeError: Cannot read property 'transform' of undefined
    at Camera.update (camera.js?a778:37)
    at E.render (babylon.js?98a3:1)
    at BabylonLayer.render (BabylonCustom.js?ef62:60)
    at Object.custom (mapbox-gl.js?d5ed:33)
    at ro.renderLayer (mapbox-gl.js?d5ed:33)
    at ro.render (mapbox-gl.js?d5ed:33)
    at r._render (mapbox-gl.js?d5ed:33)
    at eval (mapbox-gl.js?d5ed:33)" ,  the   matrix has no the transform function? please help me !

**Comment by peterqliu** [comment]
This thread has drifted a bit, but given [the lack of consensus](https://github.com/mapbox/mapbox-gl-js/issues/7395#issuecomment-428683379) around what's most useful for camera settings, I'm inclined to funnel more development toward threebox and similar plugins. It feels like the most expedient way to solve this challenge in a standardized way, and get developers up and running with custom layers.

closing for now, barring better ideas

**Comment by danvk** [comment]
@indus do you have a complete example of that `render()` method working? I like the approach but I'm having trouble getting it to work in practice.

Context is this Stack Overflow question (if others on this thread have ideas, please answer it!)
[Raycast in Three.js with only a projection matrix](https://stackoverflow.com/questions/59163141/raycast-in-three-js-with-only-a-projection-matrix?noredirect=1#comment106841846_59163141)

**Comment by isolin** [comment]
@danvk Answered :) [TL;DR - a full working example](https://jsfiddle.net/5vpL7ays/6/). After messing around a bit with coordinate systems, it tracks object picking and returns distance in meters!

Thanks to @markusjohnsson for inspiration!

**Comment by sknightq** [comment]
> @danvk Answered :) [TL;DR - a full working example](https://jsfiddle.net/5vpL7ays/7/). After messing around a bit with coordinate systems, it tracks object picking and returns distance in meters!
> 
> Thanks to @markusjohnsson for inspiration!

If someone using the Threejs version is above v0.117.1. Don't forget to change some code in `render()` function.
from 
`this.renderer.state.reset();`
to
`this.renderer.resetState();`

[Detail](https://github.com/mrdoob/three.js/issues/22407)
